intermittent_tracker/static/index.html

```diff
@@ -0,0 +1,292 @@
+            <td v-if="columnVisible('message')"><pre class="ugc">{{ attempt.message }}</pre></td>
+            <td v-if="columnVisible('stack')"><pre class="ugc">{{ attempt.stack }}</pre></td>
+            <td v-if="columnVisible('branch')"><a :href="filter('branch', attempt)">{{ attempt.branch }}</a></td>
+            <td v-if="columnVisible('pull_url')"><a :href="filter('pull_url', attempt)">{{ short(attempt.pull_url) }}</a></td>
+            <td v-if="columnVisible('build_url')"><a :href="filter('build_url', attempt)">{{ short(attempt.build_url) }}</a></td>
+        </tr>
+    </table>
+</script>
+<script type="text/x-template" id="Time">
+    <time v-if="relativeTo != null" :datetime="iso" :title="iso">{{ relativeText }}</time>
+    <time v-else :datetime="iso" :title="iso">{{ absoluteText }}</time>
+</script>
+<script type="text/x-template" id="Detail">
+    <a class="ugc" v-if="field.endsWith('_url')" :href="value" target="_blank">{{ value }}</a>
+    <pre class="ugc" v-else-if="value != null">{{ value }}</pre>
+    <em class="ugc" v-else>(null)</em>
+</script>
+<script>
+    const ALL_KEYS = ["path", "subtest", "expected", "actual", "message", "stack", "branch", "build_url", "pull_url"];
+    const FILTER_KEYS = ["path", "subtest", "expected", "actual", "branch", "build_url", "pull_url"];
+    const BULKY_KEYS = ["expected", "message", "stack"];
+
+    const Recent = {
+        template: "#Recent",
+        data() {
+            const params = new URLSearchParams(location.search);
+            const filters = {};
+            const filterKeys = [];
+            for (const key of FILTER_KEYS) {
+                if (params.has(key)) {
+                    filters[key] = params.get(key);
+                    filterKeys.push(key);
+                }
+            }
+            return {
+                // inputs for next request
+                since: null,
+                filters,
+                filterKeys,
+
+                // last response time, for “as of”
+                lastUpdate: null,
+
+                // last response body, to optimise Recent#update()
+                lastResult: [],
+
+                // actual data to be shown
+                attempts: [],
+                attemptsRaw: [],
+                allSame: {},
+                allSameKeys: [],
+            };
+        },
+        mounted() {
+            this.update();
+        },
+        methods: {
+            async update() {
+                try {
+                    // fetch attempts where time >= since (and our filters).
+                    // we use >= rather than > because a request made early in
+                    // an epoch second might not contain results submitted later
+                    // in that second, so we need to make sure we pick them up
+                    // in the next request. this complicates our logic a little.
+                    const result = await fetchRecent(this.since, this.filters);
```

Comment:
this approach (and the rest of the update logic) assumes that the time column in the database represents the time of submission, but the dashboard api currently uses a caller-provided time, which in my test-wpt prototype has been the time each test was actually run.

backdating rows like this means they can get missed in live updates. for example, in the timeline below, after we receive results for job 1, we would set since = the time of the last unexpected result. when job 2 later submits its results with a wide variety of times, some of them (yellow) will never be sent to the client.

![image](https://user-images.githubusercontent.com/465303/217196816-082fe71f-f865-4ac4-bb3e-d44f1ec65436.png)

Reply:
This is an interesting issue. Thinking about this a bit, the actual time the test is run can be an interesting clue to solving issues related to the time of the day (or year) when they are run. For instance, if we notice that tests always flake around midnight, we might notice this pattern. In that sense, I think the run time could be pretty useful. That said, I imagine either approach is okay as submission time is likely near the actual time of the test (within 10 minutes).

Reply:
good point! fixed in dc9fac3db4354eea4ae377fbdf1ead800dc064d8, which also greatly simplifies the update logic, cuts sqlite query times from ~2 ms to ~40 µs, and cuts database size for 6 full wpt runs from 7.4 MB to 4.5 MB.

Reply:
aforementioned test data: [dashboard.sqlite.zip](https://github.com/servo/intermittent-tracker/files/10677564/dashboard.sqlite.zip) (7.4 MB in v1 schema)

Reply:
new patch for playing with test-wpt: [test-wpt-v3.patch.zip](https://github.com/servo/intermittent-tracker/files/10677997/test-wpt-v3.patch.zip) ([old](https://github.com/servo/servo/pull/29320#issuecomment-1412551323): [test-wpt-v2.patch.zip](https://github.com/servo/intermittent-tracker/files/10678004/test-wpt-v2.patch.zip))